torch/_library/opaque_object.py

```diff
@@ -1,10 +1,36 @@
 from typing import Any, NewType
 
 import torch
+import torch.utils._pytree as pytree
 
 from .fake_class_registry import register_fake_class
 
 
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# Note [Opaque Objects]
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# Opaque objects are objects that we allow to be passed into custom operators.
+#
+# There are two kinds of opaque types: VALUE types and REFERENCE types.
+# The distinction determines how torch.compile handles the object.
+#
+# REFERENCE TYPES (default):
+# - Represents mutable stateful objects and are treated as black boxes.
+# - In torch.compile, the object is treated as an input/output to the graph.
+# - Register with: register_opaque_type(MyClass)
+#
+# VALUE TYPES:
+# - Represents constant values that are embedded in the graph
+# - When torch.compiling, the object is treated as a constant and the graph
+#   specializes on the constant. Guards are based on equality (obj.__eq__),
+#   meaning torch.compile will recompile if the hash of the object changes.
+# - The type's __eq__, __hash__, and __repr__ must be implemented.
```

Comment:
This note is basically a design note, so.. the question is why do these need to be implemented and what are they used for?